### PR TITLE
Support preemptive acknacks for Connext

### PR DESCRIPTION
**Pull request by jrw972**

### CODE REVIEW

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.h [hunk 1]
```diff
@@ -606,10 +606,25 @@ class OpenDDS_Rtps_Udp_Export RtpsUdpDataLink
...
+    SequenceNumber preemptive_acknack_base() const
+    {
+      // RTI expects 0 while the spec implies it should be 1.
+      return (id_.guidPrefix[0] == 0x01 && id_.guidPrefix[1] == 0x01) ? 0 : 1;
+    }
+
+    CORBA::Long next_acknack_count()
+    {
+      // Reflect the heartbeat count for OpenDDS.
+      return (id_.guidPrefix[0] == 0x01 && id_.guidPrefix[1] == 0x03) ? heartbeat_recvd_count_ : ++acknack_count_;
```
Comment (iguessthislldo) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.h:623]: Could use the constant?

https://github.com/OpenDDS/OpenDDS/blob/e80a522c7ae1b60d0bec6a3691f046c86eb89d04/dds/DCPS/GuidUtils.h#L29

Could also make a function in `GuidUtils.h` that checks if `GUID_t` has a `GuidVendorId_t`.

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp [hunk 2]
```diff
@@ -1662,7 +1662,8 @@ RtpsUdpDataLink::RtpsReader::RtpsReader(const RtpsUdpDataLink_rch& link, const G
   , stopping_(false)
   , nackfrag_count_(0)
   , preassociation_task_(make_rch<SporadicEvent>(link->event_dispatcher(), make_rch<PmfNowEvent<RtpsReader> >(rchandle_from(this), &RtpsUdpDataLink::RtpsReader::send_preassociation_acknacks)))
-  , heartbeat_period_(link ? link->config()->heartbeat_period() : TimeDuration(RtpsUdpInst::DEFAULT_HEARTBEAT_PERIOD_SEC))
+  , initial_fallback_(link ? link->config()->heartbeat_period() : TimeDuration(RtpsUdpInst::DEFAULT_HEARTBEAT_PERIOD_SEC))
+ , fallback_(initial_fallback_)
```
Comment (iguessthislldo) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp:1666]:
```suggestion
  , fallback_(initial_fallback_)
```

**Review by iguessthislldo — CHANGES_REQUESTED** [review]

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.h [hunk 3]
```diff
@@ -181,6 +184,7 @@ class OpenDDS_Rtps_Udp_Export RtpsUdpDataLink
   bool associated(const GUID_t& local, const GUID_t& remote,
                   bool local_reliable, bool remote_reliable,
                   bool local_durable, bool remote_durable,
+                  const RTPS::VendorId_t vendor_id,
```
Comment (mitza-oci) [inline dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.h:187]:
```suggestion
                  RTPS::VendorId_t vendor_id,
```

Or pass by `const &`

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.h (hunk 3 above)
Comment (jrw972) [reply dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.h:187]: I think I used const ref everywhere else so I wonder why I didn't here.

File: docs/news.d/preemptive_acknacks.rst [hunk 4]
```diff
@@ -0,0 +1,6 @@
+.. news-prs: 4735
+
+.. news-start-section: Additions
+- OpenDDS now sends preemptive (preassociation) acknacks in a form that RTI Connext understands.
+- Preassociation acknacks now falloff like heartbeats.
```
Comment (mitza-oci) [inline docs/news.d/preemptive_acknacks.rst:5]:
```suggestion
- Preassociation acknacks now fall off like heartbeats.
```

Should Preassociation be Pre-association?

Comment (jrw972) [reply docs/news.d/preemptive_acknacks.rst]: I think so.

**Review by mitza-oci — APPROVED** [review]

File: dds/DCPS/RTPS/Spdp.cpp [hunk 5]
```diff
@@ -4679,6 +4679,23 @@ DDS::ParticipantBuiltinTopicData Spdp::get_part_bit_data(bool secure) const
...
+  return get_vendor_id_i(guid);
+}
+
+VendorId_t Spdp::get_vendor_id_i(const GUID_t& guid) const
+{
+  const VendorId_t unknown_vendor = { 0, 0 };
+  DiscoveredParticipantConstIter iter = participants_.find(make_part_guid(guid));
+  if (iter != participants_.end()) {
+    return iter->second.pdata_.participantProxy.vendorId;
+  }
```
Comment (iguessthislldo) [inline dds/DCPS/RTPS/Spdp.cpp:4695]: The vendor id isn't in `guid`?

Comment (mitza-oci) [reply dds/DCPS/RTPS/Spdp.cpp:4695]: GUID is not guaranteed to include a VendorId.  The RTPS spec requires and and the DDS Security spec (with built-in plugins) effectively forbids it.  Even before DDS Security, real world experience with non-OpenDDS implementations indicated that they did not follow this part of the RTPS spec.

**Review by iguessthislldo — APPROVED** [review]

**Pull request merged**